test-cursors/Test/Database/LMDB/Simple/Cursor/Lockstep.hs

```diff
@@ -0,0 +1,503 @@
+    CursorPut pfMay k v -> fmap Some $
+      CursorPut <$> shrinkCPutFlag pfMay <*> shrinkKey k <*> shrinkValue v
+    CursorDel dfMay     -> fmap Some $ CursorDel <$> shrinkCDelFlag dfMay
+  where
+    shrinkOp :: MDB_cursor_op -> [MDB_cursor_op]
+    shrinkOp MDB_GET_CURRENT = []
+    shrinkOp _               = [MDB_GET_CURRENT]
+
+    shrinkCPutFlag :: Maybe CPutFlag -> [Maybe CPutFlag]
+    shrinkCPutFlag Nothing = []
+    shrinkCPutFlag _       = [Nothing]
+
+    shrinkCDelFlag :: Maybe CDelFlag -> [Maybe CDelFlag]
+    shrinkCDelFlag Nothing = []
+    shrinkCDelFlag _       = [Nothing]
+
+    shrinkKey :: k -> [k]
+    shrinkKey = QC.shrink
+
+    shrinkValue :: v -> [v]
+    shrinkValue = QC.shrink
+
+{-------------------------------------------------------------------------------
+  Interpret @'Op'@ against @'ModelValue'@
+-------------------------------------------------------------------------------}
+
+instance InterpretOp Op (ModelValue (CursorState k v)) where
+  intOp OpId         = Just
+  intOp OpFst        = \case MPair   x -> Just (fst x)
+  intOp OpSnd        = \case MPair   x -> Just (snd x)
+  intOp OpLeft       = \case MEither x -> either Just (const Nothing) x
+  intOp OpRight      = \case MEither x -> either (const Nothing) Just x
+  intOp (OpComp g f) = intOp g <=< intOp f
+
+{-------------------------------------------------------------------------------
+  Interpreter for @'CursorM'@
+-------------------------------------------------------------------------------}
+
+runCM ::
+     (Serialise k, Serialise v)
+  => LockstepAction (CursorState k v) a
+  -> LookUp (RealMonad k v)
+  -> RealMonad k v (Realized (RealMonad k v) a)
+runCM act _lookUp = case act of
+  CursorGet op        -> catchErr $ cgetG op
+  CursorGetSet k      -> catchErr $ cgetSet k
+  CursorGetSetKey k   -> catchErr $ cgetSetKey k
+  CursorGetSetRange k -> catchErr $ cgetSetRange k
+  CursorPut fls k v   -> catchErr $ cputG fls k v
+  CursorDel fls       -> catchErr $ cdelG fls
+
+catchErr :: RealMonad k v a -> RealMonad k v (Either Err a)
+catchErr act = catch (Right <$> act) handler
+  where
+    handler :: LMDB_Error -> RealMonad k v (Either Err a)
+    handler e = maybe (throwM e) (return . Left) (fromLMDBError e)
+
+{-------------------------------------------------------------------------------
+  Statistics and tagging
+-------------------------------------------------------------------------------}
+
+data Stats k v = Stats
+  deriving stock (Show, Eq)
+
+initStats :: Stats k v
+initStats = Stats
+
+updateStats ::
+     Ord k
+  => LockstepAction (CursorState k v) a
+  -> CursorVal k v a
+  -> Stats k v
+  -> Stats k v
+updateStats _ _ stats = stats
+
+data Tag = Tag
+  deriving (Show, Eq)
+
+tagCursorAction ::
+     (Show k, Show v, Ord k)
+  => Stats k v
+  -> LockstepAction (CursorState k v) a
+  -> ModelValue (CursorState k v) a
+  -> [Tag]
+tagCursorAction _ _ _ = []
```

Comment:
The tests have no tags yet. Suggestions for sensible tags are welcome! One minimal example I can think of is a simple "round trip": putting, getting and deleting a key.